### DISCUSSION TITLE
Recursive filter

### DISCUSSION THREAD

**Discussion by DigNZ** [pocketbase/pocketbase#2413] [Q&A] [answered]
This might be a stretch!

I have a collection called 'Group' which has a name and a parent which is also a group. Given groups A, B, C
 where A is the parent of B and B is the parent of C.

Records of another collection 'Things' are assigned a group.

A user has access to a group and all the children of that group.

If a user has access to Group B, they would see Things in groups B and C but not A.
If a user has access to Group A they would see Things in groups A, B and C.

What is the best way to set the filter rules to support this given any depth of heirachy. Or is there a better way to set up the collections to support this? OR is it a case for custom hooks at the framework level?

**Comment by ganigeorgiev** [comment, marked as answer]
Hm, unfortunately I don't think there is an easy way to do this with just the API rules and will depend how the user are associated to the group

One possible workaround could be to create a helper view collection that will hold a flatten list with each group-parent relation and then join it in the main group collection to check whether the user has access to at least one of the related group-parent pairs.

Assuming that the user are associated to the group via a junction collection `userGroup`, this could look something like (_I didn't tested it and most likely you'll have to adjust it to your actual structure_):

1. Create a helper view "groupParent"
    ```sql
    WITH cteGroup(id, parent) as 
    (
        SELECT g.id, g.parent
        FROM `group` as g
        UNION ALL
        SELECT g2.id, p.parent
        FROM cteGroup AS p
        JOIN `group` as g2 on g2.parent = p.id
        WHERE p.parent != ""
    )
    SELECT
      (ROW_NUMBER() OVER()) as id,
      cteGroup.id as `group`,
      cteGroup.parent
    FROM cteGroup;
    ```

2. Example "groups" API rule

    ```js
    @request.auth.id != "" &&
    @collection.groupParent.group ?= id &&
    @collection.userGroup.group ?= @collection.groupParents.parent
    @collection.userGroup.user ?= @request.auth.id
    ```

**Reply by DigNZ** [reply]
A variation of this worked perfectly. Thanks!

**Reply by aonghas** [reply]
Hey @DigNZ can you share the solution that you had to this? I have the exact same use case and struggling to get the correct view collection query with recursive descendents!

**Comment by DigNZ** [comment]
Hi,

Just a note for the record. This stopped working in the latest version of pocketbase 0.17.4 because of the column alias 'group' which was then passed through some internal process without quotes and produced an invalid query. Pleased that pocketbase was generally robust with this (everything else kept running even though it used this view table for permissions) but I did need to change the name of the column and update the permissions logic thoughout to match. All good now.

**Reply by ganigeorgiev** [reply]
Hm, this is most likely caused by the recent View query wrapping in https://github.com/pocketbase/pocketbase/issues/3110.

I'll push a fix tomorrow.

**Reply by ganigeorgiev** [reply]
_I've updated the tests and submitted a small fix in [v0.17.5 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.17.5)._
